table/block_based/block_based_table_builder.cc

```diff
@@ -1886,10 +1897,72 @@ void BlockBasedTableBuilder::EnterUnbuffered() {
                                         r->pending_handle);
       }
     }
-
+    // TODO: figure out when we can ensure all the buffered data blocks are flushed and call this function
+    // TODO: figure out what is the right status message for successfully flushing all the buffered data blocks
+    if (ok()) {
+      FreeAllBufferedDataBlocskMemWithCache();
+    }
     std::swap(iter, next_block_iter);
   }
   r->data_block_buffers.clear();
+  r->buffer_exceeds_global_memory_limit = false;
+}
+
+// TODO figure out a better name for this function to reflect what it does in
+// the context of this class
+void BlockBasedTableBuilder::ReserveLastBufferedDataBlockMemWithCache() {
+  // TODO-t figure out a better way to represent there is no need to reserve mem due to no cache
+  CacheRep* cache_rep = rep_->cache_rep;
+  if (cache_rep == nullptr) {
+    return;
+  }
+
+  // TODO figure out whether lock is needed for the cache
+
+  size_t kSizeDummyEntry = rep_->data_block_buffers.back().size();
+  Cache::Handle* handle = nullptr;
+  // TODO-l learn how cache key works
+  // TODO-l learn how (dummy) entries are deleted in cache
+  Status status = cache_rep->cache_->Insert(
+      cache_rep->GetNextCacheKey(), nullptr, kSizeDummyEntry,
+      GetNoopDeleterForRole<
+          CacheEntryRole::kCompressionDictionaryBuildingBuffer>(),
+      &handle);
+  // TODO figure out if we still need to push back a handle when insertion fails
+  cache_rep->dummy_handles_.push_back(handle);
+
+  // TODO-q ask about how to identify the specific insertion error due to cache
+  // full other than comparing the full msg string
+  // TODO figure out if there is any other insertion error we need to pay
+  // attention to
+  if (status.IsIncomplete() &&
+      strcmp(status.getState(), "Insert failed due to LRU cache being full.")) {
```

Comment:
Question 4:
Just by tracing down to what are the possible statuses returned by insertion to cache, it seems that the status.IsIncomplete() is equivalent to "insertion failure due to cache being full". Am I correct?